Managing Authors using Github Issues Wehbook, Ngrok for testing, Rake task to populate Github issues

basilkhan05 opened the pull request:
## Set of Features

- Webhook Handler methods in authors controller handle `opened`, `closed`, and `edited` events on issues
- Add default book for author when created
- Remove all of author's books when author is deleted (issue closed)
- Rake task - populate all issues from data in DB (all authors that do not have a `github_issue_id` associated with them)
- Setup Ngrok URL with ngrok gem for webhook setup and testing
- Update documentation on setup

## Walkthrough of the development app and features in this PR

The following screencast is a playlist of 3 videos that walks through a demo and setup of the development environment.

[![](http://play.vidyard.com/4MiJyB7M68qMsQUARPeP2B.jpg)](https://watch.basilkhan.ca/watch/4MiJyB7M68qMsQUARPeP2B/playlist/two-column)
[Click here to view the screencast](https://watch.basilkhan.ca/watch/4MiJyB7M68qMsQUARPeP2B/playlist/two-column)

## Not implemented
- Automatic setup of Wehbook when running `rails s` using Github API

## Assumptions:

- There can be multiple authors with the same name
  - As a result (Github Issue ID will be used to associate authors to issues)
  - When running the rake task, the creation webhook will check for an existing author with `github_issue_id` and update it if the value of `github_issue_id` is nil
- First default book for new authors will always be called "The Journey of {{author.name}}" as an example, untill there is a mechanism to add and manage books
- All Books added from Github issues are self-published

basilkhan05 committed:
Add biography attribute to author model

basilkhan05 committed:
Add Webhook Handler methods in AuthorsController for actions
- opened
- edited
- closed

basilkhan05 committed:
Add creation and deletion of books in webhook
- Add default book when author is create (Issue opened)
- Delete all books when Authors are removed (Issue closed)

basilkhan05 committed:
Add Rake task to populate issues in Github
- Add Octokit Gem
- Gitub rake tasks with basic auth

basilkhan05 committed:
Add ngrok-tunnel gem  for testing webhooks from Github in development
- configure puma.rb file with Ngrok option

basilkhan05 committed:
refactor Github event handlers into helper module
- handle_github_issue_opened
- handle_github_issue_edited
- handle_github_issue_closed

basilkhan05 committed:
refactor set author from github_issue_id before method calls in controller

basilkhan05 committed:
update handle_github_issue_edited to edit bio_book for authors
- When the author is edited from Github issues

basilkhan05 committed:
Update documentation for setup in README

undergroundwebdesigns commented on app/controllers/authors_controller.rb:
```diff
@@ -33,6 +34,20 @@ def update
   end
 
+  # POST /authors/github_webhook
+  def github_webhook
```
As written, this endpoint will accept any input, whether it comes from Github or not. Please improve this to validate that the request came from Github and raise an error if not?

basilkhan05 replied:
Makes sense. My approach would be to use the secret token in the webhook and save that as environment variable in the application. Is this what you would recommend?

basilkhan05 replied:
I have added this and pushed this to the branch using a secret token. Let me know your thoughts.

undergroundwebdesigns commented on app/helpers/github_webhook_helper.rb:
```diff
@@ -0,0 +1,39 @@
+    
+    params.merge!(author_payload)
+    self.update
+    bio_book = @author.books.find_by(:price => 0)
```
I think you would be better off just finding the first book here, this code is brittle and could lead to data corruption as soon as an author has more than one free book. Even finding the first book isn't _incredibly_ robust, but I think it's a better match for the project specs than just finding the first free book.

basilkhan05 replied:
Agreed. The reason I implemented it this way (my fault for not including this in the assumptions) was because I assumed no books would be free, except the default one, when the author is created. Will push a fix shortly.

undergroundwebdesigns commented on app/helpers/github_webhook_helper.rb:
```diff
@@ -0,0 +1,39 @@
+  def handle_github_issue_closed(event)
+    books = Book.where(:author => @author)
+    books.destroy_all
+    self.destroy
```
Why did you choose to manually delete the author's books here, rather than using a `dependent: :destroy` relationship on the model? Using a foreign key with cascade delete would be even better, but I see you're using Sqlite, which doesn't support that :(

basilkhan05 replied:
Thanks Alex. This should be fixed as well. As to the `why` for this, I completed forgot about the `dependent: :destroy` relationship in the model and definitely was a much better and cleaner way to go.

undergroundwebdesigns commented on config/puma.rb:
```diff
@@ -15,6 +16,14 @@
+    :port => 3000
+  }
+  ngrok_url = Ngrok::Tunnel.start(options)
+  puts "=> Your Ngrok URL --> #{ngrok_url}"
```
Nice!

undergroundwebdesigns commented on config/puma.rb:
```diff
@@ -15,6 +16,14 @@
 
+if ENV['RAILS_ENV'] == 'development'
+  options = {
+    :port => 3000
```
If you look higher up in the file, Puma is using the port sourced from the `PORT` env variable, with a fall back on port 3000. Please update this line to use the same port so that ngrok still works even if I configure a different port.

```suggestion
    :port => ENV.fetch("PORT") { 3000 }
```

basilkhan05 replied:
I have refactored this section

undergroundwebdesigns commented on app/helpers/github_webhook_helper.rb:
```diff
@@ -0,0 +1,39 @@
+module GithubWebhookHelper
```
As written this helper is very tightly coupled to the controller. Generally we prefer to keep any business logic out of the controller, as it's very difficult to unit-test a controller (there's a lot of setup of complicated objects like the request, params, etc) and we've frequently found that business logic needs to be re-used in other contexts (e.g. other controllers, or even rake tasks or background jobs).

Can you refactor these helper functions out into a class / classes that are called from the controller, but could equally easily be called from a rake task or other entry point?

basilkhan05 replied:
@undergroundwebdesigns Thanks for all the amazing feedback! I have made all the changes and refactored the `github_webhook_helper` to a more decoupled `github_service`. Let me know your thoughts. The reason I decided to couple it with the controller was to use the existing controller method JSON responses. However, I agree with your point about tests as these may change over time and the helper methods may break as writing tests would be difficult for the `github_webhook` route.

undergroundwebdesigns reviewed (changes requested):
Left some initial comments and suggestions. Great first-pass at this, the code is pretty easy to follow, and the readme looks good.

NOTE: For all my comments, I'm interested as much / more in _why_ you made the decisions you did as I am in getting my proposed changes in, so if you could comment back with a bit more details of your thought process around each of the areas I commented on that would be great!

basilkhan05 committed:
Change ngrok port options to change with PORT env variable
- use ENV.fetch("PORT") { 3000 } instead of hard conding port 3000

basilkhan05 committed:
refactor and decouple github_service from controller
- refactor github_webhook_helper to github_service
- add dependent: :destroy for Author's books
- refactor authors_controller

basilkhan05 committed:
add verify signature to github_webhook action
- use secret token to verify the request is coming from Github
- setup environment variable for GITHUB_WEBHOOK_SECRET_TOKEN
- use dotenv gem for env variables

basilkhan05 committed:
update README to include env variable setup
- secret token for Github Webhook in Webhook setup and in .env file

undergroundwebdesigns commented on app/services/github_service.rb:
```diff
@@ -0,0 +1,38 @@
+  def initialize(event)
+    @event = event
+    @author = Author.find_by(:github_issue_id => event["issue"]["id"])
+  end
```
This is a definite improvement! I think this class could be further improved by naming the class `GithubEventProcessor` or something similar, and giving it a `process` event, that looks at the event's `action` and decides which method to call. The individual methods can then be made private, and your calling code doesn't need to `send` anything, which reads a bit nicer.

Since this is an interview project don't worry about actually making that change this time unless you really want to ;)

basilkhan05 replied:
Thanks @undergroundwebdesigns! I just pushed this change. Is this what you had in mind?

basilkhan05 committed:
refactor github_service to github_event_processor
- add process method to consume the webhook JSON event and type and call appropriate private method
- remove github_webhook_helper
